### PR TITLE
[Feature] Checkbox Implementation

### PR DESCRIPTION
**Pull request by narin**
## Description of Change

Implementation ticket for Checkbox component ([Figma design](https://www.figma.com/design/Zzt8z60hCtdEzXx2GFWghH/branch/mAMh8vyVgsevAOungfvGN6/%F0%9F%93%90-Component-Library---Design-System---VA-Mobile?node-id=1414-454&node-type=canvas&t=XqBHXGrMGvDIo65p-0))
- Required props
  - label
  - onPress
- Optional props
  - checked
  - description
  - descriptiona11y
  - hint
  - headerA11y
  - error
  - errorA11y
  - interderminate: shows indeterminate icon (overrides checked)
  - required: displays (*Required) text next to label
  - tile

### Notes:
- **For error text:** screen reader should announce "Error", then the provided error text
- Pressable area is the checkbox icon, label, description OR the entire tiled area when `tile={true}`
- Tested in the app by returning the new Checkbox component in VASelector when selector type is Checkbox or unspecified

#### Testing Packages

components
- 0.24.2-alpha.0
- 0.24.2-alpha.1

#### iOS

https://github.com/user-attachments/assets/282ce1fc-314c-4ddf-96e8-619167d2f565

#### iOS Voiceover

https://github.com/user-attachments/assets/04d320bd-691e-4819-bc25-fd5ee34e3972

#### Android

https://github.com/user-attachments/assets/315ead94-3567-4125-bda6-099ff2d2ebb0

#### Android Talkback

https://github.com/user-attachments/assets/e33f77f6-3207-42bd-8ef1-5b97b5750064

#### va-mobile-app

https://github.com/user-attachments/assets/d5ed9765-2959-4b16-bf34-86533e47fcc5

#### Web

https://github.com/user-attachments/assets/f780adfb-02f3-4f6f-99c5-a2c14e5df398

## Testing

- [x] Tested on iOS 
- [x] Tested on Android 
- [x] Tested on Web

## PR Checklist
Code reviewer validation:
- General
	- [x] PR is linked to ticket(s)
	- [x] PR has `changelog` label applied if it's to be included in the changelog
	- [x] Acceptance criteria: 
		- All satisfied _or_
		- Documented reason for not being performed _or_
		- Split to separate ticket and ticket is linked by relevant AC(s)
	- [x] Above PR sections adequately filled out
	- [x] If any breaking changes, in accordance with the [pre-1.0.0 versioning guidelines](https://github.com/department-of-veterans-affairs/va-mobile-library#versioning-policy): a CU ticket has been created for the VA Mobile App detailing necessary adjustments with the package version that will be published by this ticket
- Code
	- [x] Tests are included if appropriate (or split to separate ticket)
	- [x] New functions have proper TSDoc annotations

## Publish

If changes warrant a new version [per the versioning guidelines](https://github.com/department-of-veterans-affairs/va-mobile-library#versioning-policy) and the PR is approved and ready to merge:
- [ ] Merge `main` into branch
- [ ] Merge branch to `main`
- [ ] Verify that [Check Component Integrations](https://github.com/department-of-veterans-affairs/va-mobile-library/actions/workflows/check-component-integrations.yml) workflow ran successfully
- [ ] [Publish new version](https://github.com/department-of-veterans-affairs/va-mobile-library/actions/workflows/publish.yml) after merging

### CODE REVIEW

File: packages/components/.storybook/web/preview.tsx
Comment (narin) [inline packages/components/.storybook/web/preview.tsx:27]: Unrelated to ticket, but sets default viewport to mobile in Storybook Web. Let me know if you have any issues with this.

File: packages/components/src/components/Checkbox/Checkbox.stories.tsx
Comment (narin) [inline packages/components/src/components/Checkbox/Checkbox.stories.tsx:23]: Figma URLs/docs/etc to be refined in #456

File: packages/components/.storybook/web/preview.tsx
Comment (TimRoe) [reply packages/components/.storybook/web/preview.tsx:27]: At face value, I didn't think I'd like it, but seems alright to me.

Only concern is if it might impact the iframes in the doc site. I don't think it will, but something to check once merged.

File: packages/components/src/components/Checkbox/Checkbox.tsx
Comment (TimRoe) [inline packages/components/src/components/Checkbox/Checkbox.tsx:5]:
```suggestion
import { spacing } from '@department-of-veterans-affairs/mobile-tokens'
import { useTranslation } from 'react-i18next'
import React, { FC } from 'react'
```

File: packages/components/src/components/Checkbox/Checkbox.tsx
Comment (TimRoe) [inline packages/components/src/components/Checkbox/Checkbox.tsx:73]: Very nitpicky, but might "block" things slightly better to remove the whitespace line between comment and first value within, while having the whitespace below to separate from the next block.

File: packages/components/src/components/Checkbox/Checkbox.tsx
Comment (TimRoe) [inline packages/components/src/components/Checkbox/Checkbox.tsx:65]: Don't feel super strongly, but suggest reworking this slightly to have separate "fonts" so that the typography tokens can be dropped in more easily than having fontFamily split from fontSize/lineHeight and needing to shuffle around more later.

File: packages/components/src/components/Checkbox/Checkbox.tsx
Comment (TimRoe) [inline packages/components/src/components/Checkbox/Checkbox.tsx:172]: Since it's applied directly to the text, not sure if we need the `||` or if it being undefined will automatically defer to the text; could potentially simplify here and other spots.

File: packages/components/src/components/Checkbox/Checkbox.tsx
Comment (TimRoe) [inline packages/components/src/components/Checkbox/Checkbox.tsx:191]: Could also be simplified if doing the suggested rework above about centering on fonts so this isn't overwriting props from baseTextStyle and is just the font + color.

File: packages/components/src/components/Checkbox/Checkbox.tsx
Comment (TimRoe) [inline packages/components/src/components/Checkbox/Checkbox.tsx:243]: Should this or the Pressable have `role="checkbox"`?

File: packages/components/src/components/Checkbox/Checkbox.tsx
Comment (TimRoe) [inline packages/components/src/components/Checkbox/Checkbox.tsx:251]: 1. Do we need both of these? Seems like they're doing the same thing
2. Do these need to be updated to also include logic for `intermediate` to be "mixed"? Currently if it was intermediate, the screen reader wouldn't indicate anything, right?

File: packages/components/src/components/Checkbox/Checkbox.tsx
Comment (TimRoe) [inline packages/components/src/components/Checkbox/Checkbox.tsx:121]: I didn't check, but think this will break with font scaling to maintain the checkbox centered with the text. Recommend following what we do in Alert/Link/Snackbar (including updates in `main`, but not yet your branch to depend on the font).

File: packages/components/src/components/Checkbox/Checkbox.tsx
Comment (TimRoe) [inline packages/components/src/components/Checkbox/Checkbox.tsx:126]: Might require changing `_icon` to a function to continue updating as expected, but suggest pulling out the icon name (and possibly fill) to not have a more complicated than single line ternary within the display logic.

File: packages/components/src/components/Checkbox/Checkbox.tsx
Comment (TimRoe) [inline packages/components/src/components/Checkbox/Checkbox.tsx:241]: Minor and nitpicky, but noting that the the order above not matching the order here is a little disorienting (probably more from looking at it in browser without Cmd + clicking around).

File: packages/components/src/components/Checkbox/Checkbox.tsx
Comment (TimRoe) [inline packages/components/src/components/Checkbox/Checkbox.tsx:245]: Not sure I want to see it changed, but just throwing out that it's more difficult to follow the Spacer use having it nested within the various pieces. But it also may be obnoxious to have all these separated by `_header && _hint ? <Spacer > : null`.

I definitely like the general principle of splitting out these named parts though as it helps to see the high level pertinent structure.

File: packages/components/src/components/Checkbox/Checkbox.tsx
Comment (TimRoe) [inline packages/components/src/components/Checkbox/Checkbox.tsx:14]: Reads as maybe too prescriptive documentation? Could also alter indeterminate and other things on the app level.

```suggestion
  /** OnPress logic to alter `checked` state or other behavior associated with the checkbox */
```
?

File: packages/components/src/components/Checkbox/Checkbox.tsx
Comment (TimRoe) [inline packages/components/src/components/Checkbox/Checkbox.tsx:21]: With how much a11y label overriding we have going here, I wonder if a better solution might be something like:
```
type textWithA11y = {
  text: string
  a11y: string
}
```

and typing this like:
```suggestion
  description?: string | textWithA11y
```

Would add some complexity using the values though, might be worth playing with locally and seeing if it shakes out to be beneficial to simplify the props from being so cluttered with text + a11y combos.

File: packages/components/src/components/Checkbox/Checkbox.tsx
Comment (TimRoe) [inline packages/components/src/components/Checkbox/Checkbox.tsx:26]:
```suggestion
  /** Optional error text. Applies error styling to checkbox if present */
```

File: packages/components/src/components/Checkbox/Checkbox.tsx
Comment (TimRoe) [inline packages/components/src/components/Checkbox/Checkbox.tsx:38]: "Optional required" feels weird to read even if it makes sense with context.

```suggestion
  /** True to append (*Required) suffix to label */
```

File: packages/components/src/components/Checkbox/Checkbox.tsx
Comment (TimRoe) [inline packages/components/src/components/Checkbox/Checkbox.tsx:42]: Feel like we probably lack consistency around it and it's just a preference thing, but just noting that the "Optional" nature of these props is covered by the typing and they could be more direct in just saying what the string does or "True to ..." for booleans.

File: packages/components/src/components/Checkbox/Checkbox.tsx
Comment (TimRoe) [inline packages/components/src/components/Checkbox/Checkbox.tsx:136]: I glanced at Figma, but is it a bug that the `indeterminate` status is styling to be gray instead of blue? Seems like it, but can't point to anything definite.

```suggestion
          checked || indeterminate
            ? theme.vadsColorFormsForegroundActive
            : theme.vadsColorFormsBorderDefault
```
?

**Review by TimRoe — CHANGES_REQUESTED** [review]
Marking request changes to denote end of first pass (besides the stories, which should be fine and have a separate ticket).

Looks mostly good, quite a few comments/suggestions for improvement. I think the notable needed fixes are:
- Indeterminate icon color bugged
- Icon alignment with text is likely bugged (didn't verify) with font scaling, should probably mimic Alert/Link/Snackbar solution
   - Wonder if this is reaching prevalence that we should have a function in the Icon file that takes the appropriate text lineHeight it needs to align with and is a function that returns the style
      - Or maybe even a component? IconAlignedWithText or something that returns the whole View and takes props for IconProps, text size it's aligned with, and any other custom View props?
      - Can probably split to a separate ticket if it makes sense to you

File: packages/components/src/components/Checkbox/Checkbox.stories.tsx
Comment (TimRoe) [inline packages/components/src/components/Checkbox/Checkbox.stories.tsx:59]: Wonder if doing the same with indeterminate will fix the weirdness observed in the one recording you had turning it on.

**Comment by narin** [comment]
- Updated prop comments and removed redundant use of “Optional”
- Separated out typography styles instead of using baseTextStyle to prepare for typography tokens
- Adjusted comment whitespace and sorted sections by display order
- Fixed icon styling to maintain centered style for font scaling
- Extracted icon prop logic from display logic
- Simplified `aria-label`
- Removed `aria-checked` since `accessibilityState` serves the same purpose.
  - Added logic for indeterminate to set as `’mixed'`

File: packages/components/src/components/Checkbox/Checkbox.tsx
Comment (TimRoe) [reply packages/components/src/components/Checkbox/Checkbox.tsx]: > I think this component probably has some of the most individual text props we'll encounter and it may be okay as is.

Seems like Radio will probably be the same and other form components will have at least 2 (label, error) if not more.

Agree with the general notion it should be applied to existing components, but could also see a caveat of `only applies to components with more than one "text" value` which would mean just this and Alert (and technically Segmented control, but that's using arrays for the labels and corresponding a11y so it's still "1" in a sense).

I wonder if we should start making a list of items to target/consider as part of going version 1.0.0 where we can feel free to make widespread breaking changes when it's expected as opposed to the current situation where flagship is largely hopping component-to-component version with no regard to what else we're up to.

All that said, this would be mostly for our benefit and not do much for consumers so maybe it's fine, if a bit obnoxious to look at.

File: packages/components/src/components/Checkbox/Checkbox.tsx
Comment (TimRoe) [inline packages/components/src/components/Checkbox/Checkbox.tsx:103]: Could be
```suggestion
    padding: spacing.vadsSpaceSm,
```
as specific sides should override the general rule.

File: packages/components/src/components/Checkbox/Checkbox.tsx
Comment (TimRoe) [inline packages/components/src/components/Checkbox/Checkbox.tsx:206]: Should no longer be needed as the other props center it within the minHeight that matches the font lineHeight
```suggestion

```

**Comment by TimRoe** [comment]
> label has to be before icon to reference label lineHeight

Don't feel strongly that it's updated, but my suggestion to match other components was such that at the top there'd be something like:
```
const font = {fontFamily: ..., fontSize: ..., lineHeight: ...}
const fontBold = ...
const fontHint = ...
```
where when we had the typography tokens that entire block could be deleted and updating the `...font`, `...fontBold`, `...fontHint` to the tokens--in that way, the needed lineHeight would be defined as it will when we have the tokens.

**Comment by TimRoe** [comment]
> Removed `aria-checked` since `accessibilityState` serves the same purpose.

Doesn't matter currently, but just making a note that `aria-checked` is the newer/more web-like implementation. Not sure there's any plans to deprecate anytime soon at the RN level, but know Tim W had pushed us towards using the newer ones because I think Web Storybook issues console warnings for using the older ones.

**Review by TimRoe — APPROVED** [review]
Some additional suggestions/comment, but nothing critical. Approved.

**Pull request merged**
